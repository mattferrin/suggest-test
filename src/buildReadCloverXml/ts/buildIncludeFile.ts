import { MeowOutput } from "../../bin/ts/cli";
import { SummaryItem } from "./undefinedOrSummary";

export function buildIncludeFile(input: MeowOutput) {
  return function includeFile(file: SummaryItem | undefined) {
    const isTested =
      (file?.statements ?? 0) > (file?.uncoveredStatements ?? 0) ||
      (file?.conditionals ?? 0) > (file?.uncoveredConditionals ?? 0);
    const isIncludeTested = input.flags.includeTested && isTested;

    const isNotTested =
      (file?.statements ?? 0) === (file?.uncoveredStatements ?? 0) &&
      (file?.conditionals ?? 0) === (file?.uncoveredConditionals ?? 0);
    const isIncludeNotTested = input.flags.includeNotTested && isNotTested;

    return isIncludeTested || isIncludeNotTested;
  };
}
