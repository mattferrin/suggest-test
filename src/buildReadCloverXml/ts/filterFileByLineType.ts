import { MeowOutput } from "../../bin/ts/cli";
import { FileArray } from "./summarizedResult";
import { Untrusted } from "./Untrusted";

export function filterFileByLineType(
  input: MeowOutput,
  projectPackage: Untrusted<{
    readonly file: FileArray;
  }>
): Untrusted<FileArray> {
  if (!input.flags.includeStatements) {
    return projectPackage?.file?.filter((projectPackageFile) => {
      return projectPackageFile?.line?.some((line) => {
        return line?.$?.type === "cond";
      });
    });
  } else {
    return projectPackage?.file;
  }
}
