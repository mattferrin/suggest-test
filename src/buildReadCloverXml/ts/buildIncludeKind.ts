import { MeowOutput } from "../../bin/ts/cli";
import { SummaryItem } from "./undefinedOrSummary";

export function buildIncludeKind(input: MeowOutput) {
  return function includeKind(file: SummaryItem | undefined) {
    const isStatementKeeper =
      (file?.statements ?? 0) > 1 && (file?.uncoveredStatements ?? 0) > 0;
    const isConditionalKeeper =
      (file?.conditionals ?? 0) > 1 && (file?.uncoveredConditionals ?? 0) > 0;

    if (!input.flags.includeConditionals) {
      return isStatementKeeper;
    } else if (!input.flags.includeStatements) {
      return isConditionalKeeper;
    } else {
      return isStatementKeeper || isConditionalKeeper;
    }
  };
}
