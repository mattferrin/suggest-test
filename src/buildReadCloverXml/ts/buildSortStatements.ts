import { MeowOutput } from "../../bin/ts/cli";
import { ReadCloverXmlOutput } from "./buildReadCloverXml";
import { summarizeIntoString } from "./summarizeIntoString";

export function buildSortStatements(input: MeowOutput) {
  return function sortStatements(
    output: ReadCloverXmlOutput
  ): ReadCloverXmlOutput {
    switch (output.tag) {
      case "errors":
        return output;
      case "summary": {
        const summaryArray = output.summaryArray
          ?.filter((file) => (file?.statements ?? 0) > 1)
          ?.sort((fileA, fileB) => {
            if (
              fileB?.statements === null ||
              fileB?.statements === undefined ||
              fileA?.statements === null ||
              fileA?.statements === undefined
            ) {
              return -1;
            } else {
              return fileA.weightedStatements > fileB.weightedStatements
                ? -1
                : 1;
            }
          });

        return {
          tag: "summary",
          summaryArray,
          summaryString: summarizeIntoString(summaryArray),
        };
      }
    }
  };
}
