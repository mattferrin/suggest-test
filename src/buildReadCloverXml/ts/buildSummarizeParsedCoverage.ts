import { MeowOutput } from "../../bin/ts/cli";
import { filterFileByLineType } from "./filterFileByLineType";
import { FileArray, summarizedResult } from "./summarizedResult";
import { SummaryItem } from "./undefinedOrSummary";
import { Untrusted } from "./Untrusted";

export type Parsed = Untrusted<{
  readonly coverage: {
    readonly project: ReadonlyArray<{
      readonly package: ReadonlyArray<{
        readonly file: FileArray;
      }>;
    }>;
  };
}>;

type SummarizeParsedCoverageOutput =
  | ReadonlyArray<SummaryItem | undefined>
  | undefined;

export function buildSummarizeParsedCoverage(input: MeowOutput) {
  return function summarizeParsedCoverage(
    parsed: Parsed
  ): SummarizeParsedCoverageOutput {
    return parsed?.coverage?.project?.flatMap((project) => {
      return project?.package?.flatMap((projectPackage) => {
        const filteredProjectPackageFile = filterFileByLineType(
          input,
          projectPackage
        );

        return summarizedResult(input, filteredProjectPackageFile);
      });
    });
  };
}
