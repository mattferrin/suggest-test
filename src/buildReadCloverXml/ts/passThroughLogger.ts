/* eslint-disable functional/no-conditional-statement */
/* eslint-disable functional/no-expression-statement */
/* eslint-disable functional/no-return-void */

import { ReadCloverXmlOutput } from "./buildReadCloverXml";

export function passThroughLogger(
  summary: ReadCloverXmlOutput
): ReadCloverXmlOutput {
  switch (summary.tag) {
    case "errors":
      summary.errors.forEach((error) => {
        console.error(error);
      });
      break;
    case "summary":
      console.log(summary.summaryString);
      break;
  }

  return summary;
}
