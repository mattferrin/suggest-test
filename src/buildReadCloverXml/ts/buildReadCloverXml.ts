import fs from "fs/promises";
import { parseStringPromise } from "xml2js";
import { MeowOutput } from "../../bin/ts/cli";
import { buildSortStatements } from "./buildSortStatements";
import { buildSummarizeParsedCoverage } from "./buildSummarizeParsedCoverage";
import { passThroughLogger } from "./passThroughLogger";
import { SummaryItem, undefinedOrSummary } from "./undefinedOrSummary";

export type ReadCloverXmlOutput =
  | { readonly tag: "errors"; readonly errors: readonly Error[] }
  | {
      readonly tag: "summary";
      readonly summaryArray: ReadonlyArray<SummaryItem | undefined>;
      readonly summaryString: string | { readonly tag: "not-ready" };
    };

export function buildReadCloverXml(input: MeowOutput) {
  // eslint-disable-next-line functional/functional-parameters
  return async function readCloverXml(): Promise<ReadCloverXmlOutput> {
    return await fs
      .readFile(input.input[0] ?? "coverage/clover.xml")
      .then(parseStringPromise)
      .then(buildSummarizeParsedCoverage(input))
      .then(undefinedOrSummary)
      .then(buildSortStatements(input))
      .catch((reason): ReadCloverXmlOutput => {
        return { tag: "errors", errors: [new Error(reason)] };
      })
      .then(passThroughLogger);
  };
}
