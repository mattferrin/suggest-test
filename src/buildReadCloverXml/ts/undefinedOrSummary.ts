import { ReadCloverXmlOutput } from "./buildReadCloverXml";

export interface SummaryItem {
  readonly conditionals: number | null | undefined;
  readonly path: string | null | undefined;
  readonly statements: number | null | undefined;
  readonly uncoveredConditionals: number;
  readonly uncoveredStatements: number;
  readonly score: number;
  readonly hasUnmarkedConditional: boolean;
}

export function undefinedOrSummary(
  summary: ReadonlyArray<SummaryItem | undefined> | undefined
): ReadCloverXmlOutput {
  if (summary === undefined || summary === null) {
    return {
      tag: "errors",
      errors: [
        new Error(
          "The parsed clover.xml has an undefined or null property somewhere"
        ),
      ],
    };
  } else {
    return {
      tag: "summary",
      summaryArray: summary,
      summaryString: { tag: "not-ready" },
    };
  }
}
