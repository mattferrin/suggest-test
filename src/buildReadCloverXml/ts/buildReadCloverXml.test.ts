import fs from "fs/promises";
import xml2js from "xml2js";
import { MeowOutput } from "../../bin/ts/cli";
import { buildReadCloverXml } from "./buildReadCloverXml";
import * as buildSortStatements from "./buildSortStatements";
import * as buildSummarizeParsedCoverage from "./buildSummarizeParsedCoverage";
import * as passThroughLogger from "./passThroughLogger";
import * as undefinedOrSummary from "./undefinedOrSummary";

function snap(input: {
  stabileSortId: string;
  comment: string | null;
  input: Parameters<typeof buildReadCloverXml>[0];
  readFileResult: jest.MockResult<typeof fs.readFile>;
  buildSortStatementsResult: jest.MockResult<typeof buildSortStatements>;
}): void {
  it(input.stabileSortId, async () => {
    /** setup mocks */
    const readFileSpy = jest
      .spyOn(fs, "readFile")
      .mockImplementation((async () => {
        if (input.readFileResult.type === "throw") {
          throw input.readFileResult.value;
        } else {
          return input.readFileResult.value;
        }
      }) as any);
    const parseStringPromiseSpy = jest
      .spyOn(xml2js, "parseStringPromise")
      .mockImplementation(async () => {
        return "parseStringPromise result";
      });
    const buildSummarizeParsedCoverageSpy = jest
      .spyOn(buildSummarizeParsedCoverage, "buildSummarizeParsedCoverage")
      .mockImplementation((() => {
        return async () => "buildSummarizeParsedCoverage result";
      }) as any);
    const undefinedOrSummarySpy = jest
      .spyOn(undefinedOrSummary, "undefinedOrSummary")
      .mockImplementation((async () => {
        return "undefinedOrSummary result";
      }) as any);
    const buildSortStatementsSpy = jest
      .spyOn(buildSortStatements, "buildSortStatements")
      .mockImplementation((() => {
        return async () => {
          if (input.buildSortStatementsResult.type === "throw") {
            throw input.buildSortStatementsResult.value;
          } else {
            return input.buildSortStatementsResult.value;
          }
        };
      }) as any);
    const passThroughLoggerSpy = jest
      .spyOn(passThroughLogger, "passThroughLogger")
      .mockImplementation((passThrough: any) => {
        return passThrough;
      });

    /** unit under test */
    const readCloverXml = buildReadCloverXml(input.input);
    const result = await readCloverXml();

    /** assertions */
    expect({
      [`${input.comment ?? "TODO: comment"}`]: {
        result,
        readFileCalls: readFileSpy.mock.calls,
        parseStringPromiseCalls: parseStringPromiseSpy.mock.calls,
        buildSummarizeParsedCoverageCalls:
          buildSummarizeParsedCoverageSpy.mock.calls,
        undefinedOrSummaryCalls: undefinedOrSummarySpy.mock.calls,
        buildSortStatementsCalls: buildSortStatementsSpy.mock.calls,
        passThroughLoggerCalls: passThroughLoggerSpy.mock.calls,
      },
    }).toMatchSnapshot();
  });
}

snap({
  stabileSortId: "dacaa33f-3d63-432b-81a2-9928a564de69",
  comment: "NO errors AND calls into every step",
  input: { flags: { cloverPath: "input flags cloverPath" } } as MeowOutput,
  readFileResult: {
    type: "return",
    value: "readFile result" as any,
  },
  buildSortStatementsResult: {
    type: "return",
    value: "buildSortStatements result" as any,
  },
});

snap({
  stabileSortId: "d4863a04-0517-4a79-97f5-b8f79bbedb7e",
  comment: "throws error just before catch AND logs it",
  input: { flags: { cloverPath: "input flags cloverPath" } } as MeowOutput,
  readFileResult: {
    type: "return",
    value: "readFile result" as any,
  },
  buildSortStatementsResult: {
    type: "throw",
    value: "buildSortStatements error" as any,
  },
});

snap({
  stabileSortId: "b67e77c1-a87e-4a20-8745-2750f57d61a5",
  comment: "throws error on first call AND logs it",
  input: { flags: { cloverPath: "input flags cloverPath" } } as MeowOutput,
  readFileResult: {
    type: "throw",
    value: "readFileResult error" as any,
  },
  buildSortStatementsResult: {
    type: "return",
    value: "buildSortStatements result" as any,
  },
});
