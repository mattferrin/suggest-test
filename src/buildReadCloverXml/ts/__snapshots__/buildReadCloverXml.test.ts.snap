// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`b67e77c1-a87e-4a20-8745-2750f57d61a5 1`] = `
{
  "throws error on first call AND logs it": {
    "buildSortStatementsCalls": [
      [
        {
          "flags": {
            "cloverPath": "input flags cloverPath",
          },
        },
      ],
    ],
    "buildSummarizeParsedCoverageCalls": [
      [
        {
          "flags": {
            "cloverPath": "input flags cloverPath",
          },
        },
      ],
    ],
    "parseStringPromiseCalls": [],
    "passThroughLoggerCalls": [
      [
        {
          "errors": [
            [Error: readFileResult error],
          ],
          "tag": "errors",
        },
      ],
    ],
    "readFileCalls": [
      [
        "input flags cloverPath",
      ],
    ],
    "result": {
      "errors": [
        [Error: readFileResult error],
      ],
      "tag": "errors",
    },
    "undefinedOrSummaryCalls": [],
  },
}
`;

exports[`d4863a04-0517-4a79-97f5-b8f79bbedb7e 1`] = `
{
  "throws error just before catch AND logs it": {
    "buildSortStatementsCalls": [
      [
        {
          "flags": {
            "cloverPath": "input flags cloverPath",
          },
        },
      ],
    ],
    "buildSummarizeParsedCoverageCalls": [
      [
        {
          "flags": {
            "cloverPath": "input flags cloverPath",
          },
        },
      ],
    ],
    "parseStringPromiseCalls": [
      [
        "readFile result",
      ],
    ],
    "passThroughLoggerCalls": [
      [
        {
          "errors": [
            [Error: buildSortStatements error],
          ],
          "tag": "errors",
        },
      ],
    ],
    "readFileCalls": [
      [
        "input flags cloverPath",
      ],
    ],
    "result": {
      "errors": [
        [Error: buildSortStatements error],
      ],
      "tag": "errors",
    },
    "undefinedOrSummaryCalls": [
      [
        "buildSummarizeParsedCoverage result",
      ],
    ],
  },
}
`;

exports[`dacaa33f-3d63-432b-81a2-9928a564de69 1`] = `
{
  "NO errors AND calls into every step": {
    "buildSortStatementsCalls": [
      [
        {
          "flags": {
            "cloverPath": "input flags cloverPath",
          },
        },
      ],
    ],
    "buildSummarizeParsedCoverageCalls": [
      [
        {
          "flags": {
            "cloverPath": "input flags cloverPath",
          },
        },
      ],
    ],
    "parseStringPromiseCalls": [
      [
        "readFile result",
      ],
    ],
    "passThroughLoggerCalls": [
      [
        "buildSortStatements result",
      ],
    ],
    "readFileCalls": [
      [
        "input flags cloverPath",
      ],
    ],
    "result": "buildSortStatements result",
    "undefinedOrSummaryCalls": [
      [
        "buildSummarizeParsedCoverage result",
      ],
    ],
  },
}
`;
